### PR TITLE
Add feature to handle field references, resolves #75

### LINKED ISSUE #75: Feature: Add support for basic references in autotype
In the source keypassx github, there is a pull request for adding some of the basic keepass references into autotype.  Is it possible to get that PR merged into this fork?
https://github.com/keepassx/keepassx/pull/184

## Expected Behavior
On Amazon.ca, I should be able to autotype/fill from ChromeIPass the records from amazon.com, as the u/p combo is stored as {REF:U@I:AMAZON.COM_UUID} and {REF:P@I:AMAZON.COM_UUID}

## Current Behavior
It sends the strings (ie: {REF:U@I:AMAZON.COM_UUID}) instead of the u/p specified in the amazon.com reference

## Possible Solution
Copy the pull request from the keepassx github and merge it in

## Context
Amazon.com and Amazon.ca both use the same credentials, but as they are using different URLs, they are seen as different sites to Keepass/KeepassX and ChromeIPass.  Keepass on Windows manages this by allowing you to create 2+ separate entries, one containing the real values, the other containing a reference via UUID to the first one (http://keepass.info/help/base/fieldrefs.html).

### PR DESCRIPTION
**Pull request by mrrsm**
## Description
  - Create popup for clone options
  - Add ability to resolve references for title, username, password, url, notes, and uuid fields when using autotype, http, and/or copying

## Motivation and Context
Allows references to be used across all fields for autotype, http, and copying
Resolves #75 

## How Has This Been Tested?
Created multiple clones of different entries
Tested http by removing the url from the original and having chromeIPass fill the appropriate data
Tested inline copy/paste by manually editing all supported fields and using the context menu to copy the values
Tested autotype similar to http by cloning a working entry and removing the association from the original

## Screenshots (if appropriate):
May/or may not be helpful but this shows the different states of the clone process

ysyunrq

jg9vihl

zqtzixx

## Types of changes
- ✅ New feature (non-breaking change which adds functionality)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document.
- ✅ My code follows the code style of this project.
- ✅ All new and existing tests passed.

### CODE REVIEW

**Comment by phoerious** [comment]
I added the screenshots directly to your post.

**Comment by TheZ3ro** [comment]
Travis gui-tests are failing

```
Running tests...
Test project /home/travis/build/keepassxreboot/keepassxc/build
    Start 22: testgui
1/2 Test #22: testgui ..........................***Failed   10.98 sec
********* Start testing of TestGui *********
Config: Using QtTest library 5.2.1, Qt 5.2.1
FAIL!  : TestGui::testCloneEntry() Compared values are not the same
   Actual   (entryView->model()->rowCount()): 1
   Expected (2)                             : 2
   Loc: [/home/travis/build/keepassxreboot/keepassxc/tests/gui/TestGui.cpp(566)]
Totals: 17 passed, 1 failed, 0 skipped
********* Finished testing of TestGui *********
```

**Comment by mrrsm** [comment]
Can someone provide some insight into how I can test a QDialog box?
The two things I have tried so far were changing the show() to an exec() for the clone entry slot but that causes the test to just hang as I can't pass in an input to the dialog box.  Leaving it as it currently is with the show() seems to have it not do anything?

**Comment by phoerious** [comment]
Generally, you use QTest::keyClick() for that.

**Comment by mrrsm** [comment]
For example in DatabaseWidget.cpp if I have the following and I run the tests I do not even see the popup window show up as far as I can tell.
```
CloneDialog* cloneDialog = new CloneDialog(this, m_db, currentEntry);
cloneDialog->show();
```
However if I still try adding the key click after the triggerAction in the unit test it still fails.
```
QTest::keyClick(m_mainWindow, Qt::Key_Enter);
```
My guess there is it is due to me sending the key press to the main window and not the dialog box which pops up.

That lead me to trying it with an exec rather then a show for the cloneDialog.  That seems to show the popup however it just freezes waiting for input.

**Comment by TheZ3ro** [comment]
@mrrsm I've made a little patch for you (since I've nothing to do :smile: )
Every dialog called from a parent can be retrived with the findChild() function.
The Widget name is declared in `src/gui/CloneDialog.ui` line 4


```
diff --git a/tests/gui/TestGui.cpp b/tests/gui/TestGui.cpp
index 0c776e0..a49590e 100644
--- a/tests/gui/TestGui.cpp
+++ b/tests/gui/TestGui.cpp
@@ -45,6 +45,7 @@
 #include "format/KeePass2Reader.h"
 #include "gui/DatabaseTabWidget.h"
 #include "gui/DatabaseWidget.h"
+#include "gui/CloneDialog.h"
 #include "gui/FileDialog.h"
 #include "gui/MainWindow.h"
 #include "gui/MessageBox.h"
@@ -563,6 +564,10 @@ void TestGui::testCloneEntry()
 
     triggerAction("actionEntryClone");
 
+    CloneDialog* cloneDialog = m_dbWidget->findChild<CloneDialog*>("CloneDialog");
+    QDialogButtonBox* cloneButtonBox = cloneDialog->findChild<QDialogButtonBox*>("buttonBox");
+    QTest::mouseClick(cloneButtonBox->button(QDialogButtonBox::Ok), Qt::LeftButton);
+
     QCOMPARE(entryView->model()->rowCount(), 2);
     Entry* entryClone = entryView->entryFromIndex(entryView->model()->index(1, 1));
     QVERIFY(entryOrg->uuid() != entryClone->uuid());
```

**Comment by mrrsm** [comment]
@TheZ3ro Thanks, I will give this another go tonight and hopefully get the tests working again

**Comment by TheZ3ro** [comment]
@mrrsm with that patch tests sould be good. Since there are conflict with DatabaseWidget,cpp can you rebase your changes on the current develop? Thanks

**Comment by mrrsm** [comment]
Thanks for the patch @TheZ3ro 
I was thrown for a loop as I was trying to do that off of the main window rather than the db widget and getting nowhere.

**Comment by droidmonkey** [comment]
Great work @mrrsm

File: src/core/Entry.cpp [hunk 1]
```diff
@@ -494,6 +494,18 @@ Entry* Entry::clone(CloneFlags flags) const
...
     entry->m_attachments->copyDataFrom(m_attachments);
+
+    if (flags & CloneUserAsRef) {
+        // Build the username refrence
+        QString username = "{REF:U@I:" + m_uuid.toHex() + "}";
+        entry->m_attributes->set(EntryAttributes::UserNameKey, username.toUpper(), m_attributes->isProtected(EntryAttributes::UserNameKey));
+    }
+
+    if (flags & ClonePassAsRef) {
+        QString password = "{REF:P@I:" + m_uuid.toHex() + "}";
```
Comment (ftes) [inline src/core/Entry.cpp]: Problem: `toHex()` seems to give upper case UUID. If the UUID of the source entry is lowercase, `resolvePlaceholder` will fail.

Comment (Darksider3) [inline src/core/Entry.cpp]: QString username = "{REF:U@I:" + m_uuid.toHex() + "}";
+        entry->m_attributes->set(EntryAttributes::UserNameKey, username.toUpper(), m_attributes->isProtected(EntryAttributes::UserNameKey));

Am i misinterpreting that, or do u simply uppercase the `username` at line 501? Just remove it?

**Comment by JetUni** [comment]
Where is the option to add a reference to an existing entry? I can't seem to find it anywhere

**Comment by mrrsm** [comment]
@JetUni There currently isn't a way via the UI to copy the reference of another entry from the entry you want the reference on.  You can get the uuid from the properties section on the entry you want to reference and then manually create the reference link on the new entry.

The other option is to clone an existing entry and check the "Replace username and password with references" option.

**Pull request merged**